vllm/config/vllm.py

```diff
@@ -296,6 +399,41 @@ def with_hf_config(
 
         return replace(self, model_config=model_config)
 
+    def _set_config_default(self, config_obj: Any, key: str, value: Any) -> None:
+        """Set config attribute to default if not already set by user.
+
+        Args:
+            config_obj: Configuration object to update.
+            key: Attribute name.
+            value: Default value (static or callable).
+        """
+        if getattr(config_obj, key) is None:
+            setattr(config_obj, key, value(self) if callable(value) else value)
+
+    def _apply_optimization_level_defaults(self, defaults: dict[str, Any]) -> None:
+        """Apply optimization level defaults using self as root.
+
+        Recursively applies values from defaults into nested config objects.
+        Only fields present in defaults are overwritten.
+
+        Args:
+            defaults: Dictionary of default values to apply.
```

Comment:
Could you explain (in a comment too) how the overwriting works? It was not clear to me what is going on from reading the code.

If the user manually specifies cudagraphs_mode = FULL_AND_PIECEWISE and does -O1, then I would expect cudagraphs_mode = FULL_AND_PIECEWISE to win out. How does that actually happen?

The usual implementation of this that I would expect is that CompilationConfig would have None defaults on the fields, and if any of them are still None then the user optimization level picks the right default for them. But I don't think this is what is going on.